src/lib/sql/sql_translator.cpp

```diff
@@ -1799,16 +1803,33 @@ std::shared_ptr<AbstractExpression> SQLTranslator::_translate_hsql_expr(
 
     case hsql::kExprCast: {
       const auto source_data_type = left->data_type();
-      if (expr.columnType.data_type == hsql::DataType::DATE) {
-        AssertInput(source_data_type == DataType::String, "Cannot cast " + left->as_column_name() + " as Date");
-        // We do not know if an expression to be casted other than a ValueExpression actually contains dates, and we
-        // cannot check this later due to the lack of a Date data type
-        AssertInput(left->type == ExpressionType::Value, "Only ValueExpressions can be casted as Date");
-        const auto date_string = boost::get<pmr_string>(static_cast<ValueExpression&>(*left).value);
-        const auto date = string_to_date(std::string{date_string});
-        if (date) return std::const_pointer_cast<AbstractExpression>(left);
-        FailInput("'" + std::string{date_string} + "' is not a valid date");
+      const auto target_hsql_data_type = expr.columnType.data_type;
+
+      if (target_hsql_data_type == hsql::DataType::DATE || target_hsql_data_type == hsql::DataType::DATETIME) {
+        AssertInput(source_data_type == DataType::String,
+                    "Cannot cast " + left->as_column_name() + " as " +
+                        std::string{magic_enum::enum_name(target_hsql_data_type)});
+        // We do not know if an expression to be casted other than a ValueExpression actually contains date time
+        // values, and we cannot check this later due to the lack of proper data types.
+        AssertInput(left->type == ExpressionType::Value, "Only ValueExpressions can be casted as " +
+                                                             std::string{magic_enum::enum_name(target_hsql_data_type)});
+        const auto input_string = std::string{boost::get<pmr_string>(static_cast<ValueExpression&>(*left).value)};
+
+        if (target_hsql_data_type == hsql::DataType::DATE) {
+          const auto date = string_to_date(input_string);
+          AssertInput(date, "'" + input_string + "' is not a valid date");
```

Comment:
"Cannot be casted"? Not that is matters much, but there a sooo many date formats (imagine these wild people from the west which put the month in front) that it might fail because we cannot cast it rather than the string not being a valid date.

Reply:
Sounds good.